walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -39,8 +40,85 @@
       return combinedTimes;
     }
 
-    List<TimeRange> requiredTimes = getGoodTimes(events, attendeesRequired, meetingDuration);
-    return requiredTimes;
+    // optimization algorithm
+    // start with one less than max number of optional attendees, and if no times work it will try
+    // one less
+    int numberOptional = Math.max(0, attendeesOptional.size() - 1);
+
+    List<TimeRange> optimalTime =
+        optimizeTime(numberOptional, events, attendeesRequired, attendeesOptional, meetingDuration);
+    return optimalTime;
+  }
+
+  /**
+   * Function that tests all possible combinations of a given number of optional attendees
+   * (numOptional) and returns the times that work. If no times work, it recursively calls itself
+   * with a smaller number of optional attendees until there are times that work. If times don't
+   * work with any optional attendees, it will return the times for only the required attendees.
+   */
+  private static List<TimeRange> optimizeTime(
+      int numOptional,
+      Collection<Event> events,
+      Collection<String> attendeesRequired,
+      Collection<String> attendeesOptionalCollection,
+      int meetingDuration) {
+
+    // change optional attendee collection into list
```

Comment:
supernit: Google style guide generally suggests proper grammar in all comments, including Capitalizing the start of each comment and adding a '.' at the end. I do find it makes code easier to read, but haven't (and won't) be making a point of reviewing this in the interest in time. Something for you to consider in future code (no need to go back and fix all the comments for this heh)

Reply:
Thanks for letting me know, I'll definitely try to do this in future code!